extra_data/components.py

```diff
@@ -330,6 +330,45 @@ def get_array(self, key, pulses=by_index[:]):
 
         return xarray.concat(arrays, pd.Index(modnos, name='module'))
 
+    def get_dask_array(self, key):
+        """Get a labelled Dask array of detector data
+
+        Dask does lazy, parallelised computing, and can work with large data
+        volumes. This method doesn't immediately load the data: that only
+        happens once you trigger a computation.
+
+        Parameters
+        ----------
+
+        key: str
+          The data to get, e.g. 'image.data' for pixel values.
+        """
+        arrays = []
+        modnos = []
+        for modno, source in sorted(self.modno_to_source.items()):
+            modnos.append(modno)
+            mod_arr = self.data.get_dask_array(source, key, labelled=True)
+
+            # At present, all the per-pulse data is stored in the 'image' key.
+            # If that changes, this check will need to change as well.
+            if key.startswith('image.'):
+                # Add pulse IDs to create multi-level index
+                pulse_id = self.data.get_array(source, 'image.pulseId')
```

Comment:
Hum, that's again tricky here :)

Usual message from Detector people is that the `pulseId` from the detector data is not a reliable information.

First, is relies on the fact that the configuration is correct (machine rate information relatively to the acquisition rate), which is not automatically set (human action).

Then, (I take AGIPD as example, because I know that one best): the first memory cell isn't working properly, and is then not used. The consequence is that it is configured to effectively store the first xray pulse to the second memory cell. Of course the `pulseId` information will not reflect that. So, let's say that the machine is operating at 4MHz and the detector at 1MHz: the actual first pulse will be labeled `4` instead of `0` (could be even worse if the configuration as mentioned in the first point is wrong).

Could we use `cellId` instead of `pulseId` here? It does not say anything with pulse correlation, but at least the information is correct. Right now, cell ids are always increasing from 0 to n for a train. I future it might have any kind of pattern I believe.

Reply:
Hmm, that's a pain.

IIRC, Steffen told me in the past that cell ID was just about detector hardware, and shouldn't be used to identify frames for scientific purposes. Pulse ID is the one that's *supposed* to be scientifically meaningful, even if there are operational problems with it.

If pulse ID is consistently wrong across a run, I think I favour 1 - i.e. if averaging across all frames with pulseId=4 makes sense, even though that's not actually pulse 4. Making it more visible would create pressure to make it more correct. Or we could let the user specify whether they want pulseId or cellId.